### PR TITLE
feat: Supporting Audio8_TTS models

### PR DESCRIPTION
**Pull request by jasonchen31**
## Summary

Native `audio.cpp` port of **Audio8 TTS Preview** as community family `audio8_tts` (DualAR: slow semantic → fast codebook → 44.1 kHz codec), derived from Fish Audio S2 Pro patterns. Ships as standalone GGUFs with no Python dependency.

- **0.6B Qwen** (`slow_backbone=qwen`, 24L, dim 896, 14H/2KV, RoPE 1e6, packed QKV+bias) — **fully native and CPU-verified** via SenseVoice ASR round-trip.
- **0.1B Falcon-H1** (`slow_backbone=falcon_h1`, dim 512, `d_inner 768=32×24`, `d_state 64`, `d_conv 4`, `dt_rank 24`, `n_group 1`, `GQA 8/2`, RoPE 1e11, `embedding_multiplier 0.1088`/`lm_head 0.0781`) — weight-complete and builds natively (GGUF `slow.embed_tokens` + `24× mamba/attention` + `semantic_output 4097`), slow AR currently a **documented stub** pending full Mamba2 port (see `docs/FALCON_H1_0.1B_PORT_PLAN.md`).

All 0.6B generation paths use ggml-native graphs; 0.1B not yet supported.

---

### 1. Community family `audio8_tts`
- **Spec:** `model_specs/audio8_tts.json` (schema-v1, languages `yue/zh/nl/en/fr/de/it/ja/ko/pl/es/auto`, tasks `tts`/`clon`, packages `audio8_tts_preview_0_6b_q8_0`/`audio8_tts_preview_0_1b_q8_0`)
- **Docs:** `docs/community_models/audio8_tts.md` (usage, architecture, GGUF packaging, validation notes), `docs/community_models/models.md` update, `docs/FALCON_H1_0.1B_PORT_PLAN.md` (M0–M4, 3–3.5d plan)
- **Tools:** `tools/community_models/convert_audio8_tts_codec.py` (zipfile/pickle `codec.pth` → `codec.safetensors`, no torch) + `tools/community_models/convert_audio8_tts.py` (snapshot → standalone GGUF, embedded config/tokenizer/spec)

### 2. Runtime (`src/community_models/audio8_tts/`)
- `types.h` / `assets.cpp` — flat ArkTTS config parsing including Falcon multipliers (embedding/lm_head/ssm/attn/mlp)
- `prompt_builder.*` — `ArkttsProcessor._prompt_segments` chat template, reference-code placement, `clean_text`
- `ar.cpp` — slow prefill/step + fast codebook graphs, embeddings (plain sum, no `1/sqrt(n)`), RAS/top-k/top-p/Gumbel sampling
  - **0.6B path:** `TransformerKVCache` + `QwenCausalDecoder` (prefill writes into step cache, `BF16` activation casts)
  - **0.1B path:** native `BackendWeightStore` load (`slow.embed_tokens 69633×512`, `in_proj 1688×512`, `conv1d 896×1×4`, `dt_bias/A_log/D`, `out_proj`, `q/k/v/o`, layernorms, `semantic_output 4097`) + **stub** `falcon_forward_stateless:861` (`TODO(Falcon-H1)` — RMSNorm + `in_proj` split `gate 768/xBC 896` + conv bias SiLU + gated `out_proj` + FFN, `attn_out=0`, no `ggml_ssm_conv/B/C/dt/A/D/ggml_ssm_scan` nor recurrent `conv[3,896]/ssm[64,32,24]` state, `O(N²)` full recompute, `expand_compact` `1024+EOS→4097`). Full Mamba2 tracked as `mamba-base.cpp:151`/`falcon-h1.cpp:132` reuse of `external/ggml` SSM.
- `codec.*` / `generator.*` / `session.*` / `tokenizer_text.*` — window-transformer codec (encode/decode), generation loop, session options (`weight_type`, `codec_weight_type`, `mem_saver`, `reference_cache_slots`)

### 3. WebUI
- `webui/configs/models_catalog.json` + `model_params.json` — `audio8-tts` / `audio8-tts-0.5b` entries (GGUF Q8, multilingual hints)
- `webui/native/dist/index.html` — merged upstream `shug0/main` via `checkout --theirs` on conflict, then **rebuilt** (`webui/native: pnpm run build`, `vite 7.3.6`, `dist/index.html` 599K) to include upstream Echo-TTS/VoxCPM/etc. + Audio8 entries.

---

### 0.6B Qwen — PASS (linux-cpu-release `audiocpp_cli`)
```bash
audiocpp_cli --model /workspace/models/Audio8-TTS-Preview-0.6b-GGUF/audio8-tts-preview-0.6b-q8_0.gguf \
  --family audio8_tts --task tts --threads 2 --text "<text>" --out /tmp/x.wav
ffmpeg -y -i /tmp/x.wav -ar 16000 -ac 1 -c:a pcm_s16le /tmp/tmp16k.wav
curl -s -X POST http://192.168.1.2:11533/v1/audio/transcriptions \
  -F file=@/tmp/tmp16k.wav -F model=sensevoice-small
```

Build: `linux-cpu-release audiocpp_cli` 100% (no regressions in `engine_model_audio8_tts`).

### 0.1B Falcon-H1 — documented stub
- Builds and runs: `2.36s` 44100Hz via `falcon_forward_stateless` stub → STT `like.` vs target (prompt-invariant logits, expected). No `/tmp` writes or Python dependency (cleaned in `04476ef`).
- Full correctness requires M2/M3 (`ggml_ssm_conv/scan` + state + hybrid attention) — plan `docs/FALCON_H1_0.1B_PORT_PLAN.md` estimates 3–3.5d; `external/ggml` kernels already present (`cpu/cuda/metal/vulkan`).

---

## Files changed (highlights)

- `src/community_models/audio8_tts/*`, `include/engine/community_models/audio8_tts/*`, `model_specs/audio8_tts.json`, `CMakeLists.txt` (`audiocpp_add_model(audio8_tts)`)
- `tools/community_models/convert_audio8_tts*.py`, `docs/community_models/audio8_tts.md`, `docs/FALCON_H1_0.1B_PORT_PLAN.md`
- `webui/configs/models_catalog.json`, `webui/native/dist/index.html` (rebuilt), plus upstream merge delta (CMake, `external/ggml` CUDA, `app/server` model memory, `docs/models/*`, `echo_tts`/`voxcpm1`/`granite5asr`)

---

### 1. Download the HF source snapshots (PyTorch preview checkpoints)

The port consumes the **official PyTorch preview checkpoints** from the Audio8 Hugging Face org — not ONNX, not pre-quantized GGUF. Each snapshot contains `config.json` / `model.safetensors` / `codec.pth` / `tokenizer.json` / `modeling_arktts.py`.

```bash
# via huggingface-cli (recommended, resume-capable)
pip install -U huggingface_hub
hf download Audio8/Audio8-TTS-Preview-0.6b --local-dir /models/Audio8-TTS-Preview-0.6b --local-dir-use-symlinks False
hf download Audio8/Audio8-TTS-Preview-0.1b --local-dir /models/Audio8-TTS-Preview-0.1b --local-dir-use-symlinks False

# or via git-lfs
git lfs clone https://huggingface.co/Audio8/Audio8-TTS-Preview-0.6b
git lfs clone https://huggingface.co/Audio8/Audio8-TTS-Preview-0.1b
```

> **File list expected in each `--model-dir`:** `config.json`, `model.safetensors` (0.6B: 1.2G, 0.1B: 324M), `codec.pth` (both 1.3G), `tokenizer.json`/`tokenizer_config.json`/`special_tokens_map.json`, `modeling_arktts.py`/`modeling_arktts_codec.py`/`processing_arktts.py`. License: Apache-2.0 — https://github.com/Audio8-AI/Audio8_TTS

If you already have a GGUF, skip to *How to test* — but the sources below are the canonical way to rebuild one.

### 2. Build the `audiocpp_gguf` converter

```bash
cmake -S . -B build --preset release  # or: cmake -S . -B build -DCMAKE_BUILD_TYPE=Release
cmake --build build --target audiocpp_gguf -j"$(nproc)"
# binary at: build/bin/audiocpp_gguf
```

### 3. Two-step offline conversion (no Python `torch` required, no downloads)

Both models use the **same two tools** — the codec converter is torch-free (zipfile/pickle + numpy + safetensors), the packager embeds 681 tensors (226 AR + 455 codec) + config/tokenizer/spec into one standalone GGUF.

#### Step 1 — Convert `codec.pth` → `codec.safetensors` (once per snapshot)

```bash
# 0.6B
python3 tools/community_models/convert_audio8_tts_codec.py \
  /models/Audio8-TTS-Preview-0.6b/codec.pth \
  /models/Audio8-TTS-Preview-0.6b/codec.safetensors

# 0.1B — same tool, different dir
python3 tools/community_models/convert_audio8_tts_codec.py \
  /models/Audio8-TTS-Preview-0.1b/codec.pth \
  /models/Audio8-TTS-Preview-0.1b/codec.safetensors
```

The converter fuses `*.parametrizations.weight.original{0,1}` and `*.weight_g/_v` weight-norm pairs into plain `*.weight` keys and validates anchors from `modeling_arktts_codec.py` (`quantizer.semantic_quantizer…`, `decoder.model.0.conv.weight`, etc.).

#### Step 2 — Package one self-contained GGUF per precision

```bash
# 0.6B — Q8_0 (default shipping precision, AR q8_0 + codec q8_0, 1.4G)
python3 tools/community_models/convert_audio8_tts.py \
  --model-dir /models/Audio8-TTS-Preview-0.6b \
  --converter build/bin/audiocpp_gguf --type q8_0 \
  --name audio8-tts-preview-0.6b --output-dir gguf-out

# 0.6B — BF16 reference (AR BF16, codec BF16, 1.8G)
python3 tools/community_models/convert_audio8_tts.py \
  --model-dir /models/Audio8-TTS-Preview-0.6b \
  --converter build/bin/audiocpp_gguf --type bf16 \
  --name audio8-tts-preview-0.6b --output-dir gguf-out

# 0.1B — Q8_0 (812M, AR q8_0 + codec q8_0)
python3 tools/community_models/convert_audio8_tts.py \
  --model-dir /models/Audio8-TTS-Preview-0.1b \
  --converter build/bin/audiocpp_gguf --type q8_0 \
  --name audio8-tts-preview-0.1b --output-dir gguf-out

# 0.1B — BF16 reference (975M)
python3 tools/community_models/convert_audio8_tts.py \
  --model-dir /models/Audio8-TTS-Preview-0.1b \
  --converter build/bin/audiocpp_gguf --type bf16 \
  --name audio8-tts-preview-0.1b --output-dir gguf-out

# If a Q8_0 package audibly drifts, keep the codec at 16-bit (conv stacks are quant-sensitive):
python3 tools/community_models/convert_audio8_tts.py --model-dir /models/Audio8-TTS-Preview-0.6b \
  --converter build/bin/audiocpp_gguf --type q8_0 --name audio8-tts-preview-0.6b-q8_0-codec-bf16
# (then pass mixed storage via --keep-type mapping in audiocpp_gguf — see tool --help)
```

**Outputs:**
- `gguf-out/audio8-tts-preview-0.6b-q8_0.gguf` (1.4G) / `-bf16.gguf` (1.8G)
- `gguf-out/audio8-tts-preview-0.1b-q8_0.gguf` (812M) / `-bf16.gguf` (975M)

Each GGUF embeds two tensor namespaces `model_weights.*` + `codec_weights.*` plus embedded `config.json`/`tokenizer.json`/`model_specs/audio8_tts.json`, so `audiocpp_cli`/`audiocpp_server` load it standalone with `--family audio8_tts`.

### 4. Load the GGUF

```bash
# 0.6B Q8
./build/bin/audiocpp_cli --family audio8_tts --task tts \
  --model gguf-out/audio8-tts-preview-0.6b-q8_0.gguf \
  --text "The quick brown fox jumps over the lazy dog." --out tts.wav --metrics

# 0.1B Q8 (build-only check until Mamba2 lands)
./build/bin/audiocpp_cli --family audio8_tts --task tts \
  --model gguf-out/audio8-tts-preview-0.1b-q8_0.gguf \
  --text "Hello world" --out tts01.wav
```

---

## TODO (follow-up PRs)

- [X] Streaming session path
- [ ] (pending until falcon-H supported by ggml) Complete 0.1B Mamba2 port (`ggml_ssm_conv/B/C/dt/A/D/scan` + ring `conv/ssm` + hybrid attention `K=1`) and SenseVoice-verify `out/*0.1b.wav`

---

## References

- Audio8 TTS: https://github.com/Audio8-AI/Audio8_TTS — `modeling_arktts.py` / `modeling_arktts_codec.py` / `processing_arktts.py` are the source of truth
- Llama SSM reference: `porting/llama.cpp/src/models/falcon-h1.cpp` + `mamba-base.cpp:151 build_mamba2_layer`

### CODE REVIEW

**Comment by 0xShug0** [comment]
@jasonchen31 Two minor issues to fix then we are good to go.

1. Can you change `max_new_tokens` to `max_tokens` in the spec json?
2.  If you set `package_defaults` to local, I don’t think the UI can handle it properly. Would you like to host the GGUFs? Or, we can clarify that we won’t host the GGUFs due to the license, as in #336 , and exclude the model from the UI.

**Comment by jasonchen31** [comment]
@0xShug0 All done. I hosted it [here](https://huggingface.co/js-byte/Audio8-TTS-Preview-0.6b-GGUF). Hope it work.
Simple question: as it uses Apache 2.0 license, it is ok to host its GGUF, right?

**Comment by 0xShug0** [comment]
> @0xShug0 All done. I hosted it [here](https://huggingface.co/js-byte/Audio8-TTS-Preview-0.6b-GGUF). Hope it work. Simple question: as it uses Apache 2.0 license, it is ok to host its GGUF, right?

Thanks! I'll test later. 

Yes it's safe to host. Interesting, I thought ASR and TTS would use the same license...

**Comment by 0xShug0** [comment]
@jasonchen31 Looks like a GitHub glitch and I can't post my comments...It's safe to host. I thought TTS would use the same CC-BY-NC-4.0 license as ASR.

**Comment by 0xShug0** [comment]
@jasonchen31 PR merged! Thanks :tada: 

For falcon-H --- It may not be worth porting it now, given that only one model would use it. But I will take a look

**Comment by jasonchen31** [comment]
Great! Thank you! Let's skip falcon-H at this moment.
